File: src/main/java/com/example/tmm022/service/PartService.java
package com.example.tmm022.service;

import com.example.tmm022.repository.PartRepository;
import com.example.tmm022.dto.PartDTO;
import com.example.tmm022.entity.Part;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PartService {

    @Autowired
    private PartRepository partRepository;

    @Autowired
    private UnitService unitService;

    public boolean checkPartExists(String partId, String unitId, String groupId, String lineId) {
        return partRepository.findPart(partId, unitId, groupId, lineId).isPresent();
    }

    public void savePart(Part part) {
        partRepository.save(part);
    }

    public void updatePart(Part part) {
        partRepository.save(part);
    }

    public List<String> lookupPartNumber(int globalParameter) {
        return partRepository.findPartNumbers(globalParameter);
    }

    public boolean validatePartNumber(String partNumber, String partDescription) {
        return partRepository.validatePartNumber(partNumber, partDescription);
    }

    public List<String> fetchLovData() {
        return unitService.fetchLovData();
    }

    public boolean validateUnitIdAndName(String unitId, String unitName) {
        return unitService.validateUnitIdAndName(unitId, unitName);
    }

    public String validateFields(PartDTO partDTO) {
        if (partDTO.getUnitId() == null || partDTO.getUnitName() == null) {
            return "Unit ID and Unit Name should not be null";
        }
        if (partDTO.getGroupId() == null || partDTO.getGroupName() == null) {
            return "Group ID and Group Name should not be null";
        }
        if (partDTO.getLineId() == null || partDTO.getLineDesc() == null) {
            return "Line ID and Line Description should not be null";
        }
        if (partDTO.getPartNo() == null || partDTO.getPartDesc() == null) {
            return "Part No and Part Description should not be null";
        }
        if (partDTO.getPartId() == null && partDTO.getGlobalParameter() == 1) {
            return "Part ID should not be null";
        }
        return null;
    }

    public boolean validatePartStatus(String partStatus) {
        return partRepository.findByPartStatus(partStatus).isPresent();
    }

    public boolean validateRequiredFields(PartDTO partDTO) {
        return partRepository.findByFields(partDTO).isPresent();
    }

    public boolean validateLineID(String lineID, String lineDesc, String unitID, String groupID) {
        return partRepository.validateLineID(lineID, lineDesc, unitID, groupID).isPresent();
    }

    public List<String> fetchLOVData(int globalParameter) {
        return partRepository.fetchLOVData(globalParameter);
    }

    public boolean validateGroupID(String groupID, String groupName, int globalParameter, String unitID) {
        return partRepository.validateGroupIDQuery(groupID, groupName, globalParameter, unitID).isPresent();
    }
}

File: src/app/services/part.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Part } from '../models/part.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class PartService {
  private apiUrl = `${environment.apiUrl}/parts`;

  constructor(private http: HttpClient) {}

  checkPartExists(partId: string, unitId: string, groupId: string, lineId: string): Observable<boolean> {
    const url = `${this.apiUrl}/exists`;
    return this.http.post<boolean>(url, { partId, unitId, groupId, lineId });
  }

  savePart(part: Part): Observable<void> {
    const url = `${this.apiUrl}/save`;
    return this.http.post<void>(url, part);
  }

  updatePart(part: Part): Observable<void> {
    const url = `${this.apiUrl}/update`;
    return this.http.put<void>(url, part);
  }

  lookupPartNumber(globalParameter: number): Observable<string[]> {
    const url = `${this.apiUrl}/lookup`;
    return this.http.get<string[]>(url, { params: { globalParameter: globalParameter.toString() } });
  }

  validatePartNumber(partNumber: string, partDescription: string): Observable<boolean> {
    const url = `${this.apiUrl}/validate`;
    return this.http.post<boolean>(url, { partNumber, partDescription });
  }

  validateFields(part: Part): string[] {
    const messages: string[] = [];
    if (!part.unitId || !part.unitName) {
      messages.push('Unit ID and Unit Name should not be null');
    }
    if (!part.groupId || !part.groupName) {
      messages.push('Group ID and Group Name should not be null');
    }
    if (!part.lineId || !part.lineDesc) {
      messages.push('Line ID and Line Description should not be null');
    }
    if (part.globalParameter === 0) {
      if (!part.partNo || !part.partDesc) {
        messages.push('Part No and Part Description should not be null');
      }
    } else if (part.globalParameter === 1) {
      if (!part.partId) {
        messages.push('Kindly choose data from LOV before changing Description');
      }
      if (!part.partNo || !part.partDesc) {
        messages.push('Part No and Part Description should not be null');
      }
    }
    return messages;
  }

  validatePartStatus(partStatus: string): Observable<boolean> {
    const url = `${this.apiUrl}/validateStatus`;
    return this.http.post<boolean>(url, { partStatus });
  }

  validateRequiredFields(fields: any): Observable<boolean> {
    const url = `${this.apiUrl}/validateFields`;
    return this.http.post<boolean>(url, fields);
  }
}
